### PR TITLE
✨ Maintenance scheduler

### PR DESCRIPTION
**Pull request by odeimaiz**
## What do these changes do?

This PR brings the first iteration on the Maintenance Scheduler.

This is how the Maintenance notification workflow goes:
- A dev team member adds a ```maintenance``` entry in the ```scheduled_maintenance``` redis database. For the time being, this entry is a string containing an object with maintenance Start and End times (UTC) and also a Reason.
- The frontend when a user logs in requests this information. It will keep requesting it hourly. If a maintenance gets scheduled, the frontend will stop requesting it once it is aware.
- This scheduled maintenance notification is added to the navigation bar and it will always be visible under the bell.
- 20' before the scheduled maintenance, a permanent Flash Message will also be shown.
- Once the frontend hits the maintenance start time, it will automatically logout the user. It will also close any study open.

##  ⚠️ devops
  - See new redis db: ```scheduled_maintenance```

maintenance

## Related issue/s

https://github.com/ITISFoundation/osparc-issues/issues/765

### CODE REVIEW

**Comment by Surfict** [comment]
Very nice, thanks !

File: services/web/server/src/simcore_service_webserver/rest_handlers.py [hunk 1]
```diff
@@ -68,3 +69,16 @@ async def get_config(request: web.Request):
...
+@routes.get(f"/{api_version_prefix}/scheduled_maintenance", name="get_scheduled_maintenance")
+async def get_scheduled_maintenance(request: web.Request):
+    """Check scheduled_maintenance table in redis"""
+
+    redis_client = get_redis_scheduled_maintenance_client(request.app)
+    hash_key = "maintenance"
+    # {"start": "2023-01-17T14:45:00.000Z", "end": "2023-01-17T23:00:00.000Z", "reason": "Release 1.0.4"}
+    maintenance_str = await redis_client.get(hash_key)
+    if maintenance_str is not None:
+        return web.json_response(data={"data": maintenance_str})
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/rest_handlers.py:83]:
```suggestion
    if maintenance_str := await redis_client.get(hash_key):
        return web.json_response(data={"data": maintenance_str})
```

**Review by sanderegg — APPROVED** [review]
Great, I made a suggestion for the backend. I think it's ok for now.

some things I have doubts about are:
- returning 204 vs 404 when there is nothing (very minor and not sure myself now)
- no validation of what is in the redis key (rest_handlers.py). How does the maintainer knows that a T is forgotten in the date-time entry? (is easy to fix with a pydantic model, and have an error logged in the webserver using something like:
```python
class _MaintenanceEntry(BaseModel):
  start: datetime
  end: datetime
  reason: str
try:
  maintenance_entry = parse_obj_as(_MaintenanceEntry, await redis_client.get(hash_key))
  return web.json_response(data={"data": jsonable_encoder(maintenance_entry)})
except ValidationError:
  logger.exception("Invalid scheduled maintenance!")

return web.json_response(status=204)
```

File: services/web/server/src/simcore_service_webserver/rest_handlers.py [hunk 2]
```diff
@@ -68,3 +69,16 @@ async def get_config(request: web.Request):
...
+
+
+@routes.get(f"/{api_version_prefix}/scheduled_maintenance", name="get_scheduled_maintenance")
+async def get_scheduled_maintenance(request: web.Request):
+    """Check scheduled_maintenance table in redis"""
+
+    redis_client = get_redis_scheduled_maintenance_client(request.app)
+    hash_key = "maintenance"
+    # {"start": "2023-01-17T14:45:00.000Z", "end": "2023-01-17T23:00:00.000Z", "reason": "Release 1.0.4"}
+    maintenance_str = await redis_client.get(hash_key)
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/rest_handlers.py:81]: I guess you need to validate the ``maintenance_str``
```python
MaintenanceData(BaseModel):
   start: datetime
   end: datetime
   reason: str

data = MaintenanceData.parse_raw(maintenance_str) 
```

**Review by pcrespov — APPROVED** [review]
Since it is an isolated entrypoint and you tested it in the front-end, I will follow up on some other details that are missing

**Review by GitHK — APPROVED** [review]
👍

**Comment by odeimaiz** [comment]
> * returning 204 vs 404 when there is nothing (very minor and not sure myself now)

I'll stick to the 204: request has succeeded, but there is nothing to return

File: services/web/server/src/simcore_service_webserver/rest_handlers.py (hunk 2 above)
Comment (odeimaiz) [reply services/web/server/src/simcore_service_webserver/rest_handlers.py]: Feel free to validate it and to some tests in the next iteration 👍

**Review by mrnicegyu11 — APPROVED** [review]
Supernice thx! Please add a devops flag + warning: we have to (Re)Name the redis DB in redis commander accordingly

**Pull request merged**